File: medium/Two-Pointers/3sum-closest/three_sum_closest.py
from typing import List

class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        """
        Find three integers whose sum is closest to target.
        Approach: Sort + Two Pointers
        Time Complexity: O(n^2)
        Space Complexity: O(log n)
        """
        nums.sort()
        n = len(nums)
        closest = nums[0] + nums[1] + nums[2]
        
        for i in range(n - 2):
            left = i + 1
            right = n - 1
            while left < right:
                total = nums[i] + nums[left] + nums[right]
                # update closest sum
                if abs(total - target) < abs(closest - target):
                    closest = total
                if total < target:
                    left += 1
                elif total > target:
                    right -= 1
                else:
                    # exact match, return immediately
                    return target
        return closest

File: medium/Two-Pointers/3sum-closest/three_sum_closest.cpp
#include <vector>
#include <algorithm>
#include <cmath>
using namespace std;

class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        /*
         * Find triplet sum closest to target.
         * Approach: Sort + Two Pointers
         * Time Complexity: O(n^2)
         * Space Complexity: O(log n)
         */
        sort(nums.begin(), nums.end());
        int n = nums.size();
        int closest = nums[0] + nums[1] + nums[2];
        
        for (int i = 0; i < n - 2; ++i) {
            int left = i + 1;
            int right = n - 1;
            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                if (abs(sum - target) < abs(closest - target)) {
                    closest = sum;
                }
                if (sum < target) {
                    left++;
                } else if (sum > target) {
                    right--;
                } else {
                    return target;
                }
            }
        }
        return closest;
    }
};
